alibi_detect/od/pytorch/base.py

```diff
@@ -0,0 +1,235 @@
+        ThresholdNotInferredError
+            Raised if threshold is not inferred.
+        """
+        if not self.threshold_inferred:
+            raise ThresholdNotInferredError(self.__class__.__name__)
+
+    @staticmethod
+    def _to_numpy(arg: Union[torch.Tensor, TorchOutlierDetectorOutput]) -> Union[np.ndarray, Dict]:
+        """Converts any `torch` tensors found in input to `numpy` arrays.
+
+        Takes a `torch` tensor or `TorchOutlierDetectorOutput` and converts any `torch` tensors found to `numpy` arrays
+
+        Parameters
+        ----------
+        x
+            Data to convert.
+
+        Returns
+        -------
+        `np.ndarray` or dictionary of containing `numpy` arrays
+        """
+        return to_numpy(arg)
+
+    def _to_tensor(self, x: Union[List, np.ndarray]) -> torch.Tensor:
+        """Converts the data to a tensor.
+
+        Parameters
+        ----------
+        x
+            Data to convert.
+        """
+        return torch.as_tensor(x, dtype=torch.float32, device=self.device)
+
+    def _ensembler(self, x: torch.Tensor) -> torch.Tensor:
+        """Aggregates and normalizes the data
+
+        If the detector has an ensembler attribute we use it to aggregate and normalize the data.
+
+        Parameters
+        ----------
+        x
+            Data to aggregate and normalize.
+
+        Returns
+        -------
+        `torch.Tensor` or original data without alteration
+
+        Raises
+        ------
+        ThresholdNotInferredError
+            If the detector is an ensemble, and the ensembler used to aggregate the outlier scores has a fittable
+            component, then the detector threshold must be inferred before predictions can be made. This is because
+            while the scoring functionality of the detector is fit within the `.fit` method on the training data
+            the ensembler has to be fit on the validation data along with the threshold and this is done in the
+            `.infer_threshold` method.
+        """
+        if hasattr(self, 'ensembler') and self.ensembler is not None:
+            # `type: ignore` here because self.ensembler here causes an error with mypy when using torch.jit.script.
+            # For some reason it thinks self.ensembler is a torch.Tensor and therefore is not callable.
+            if not torch.jit.is_scripting():
+                if not self.ensembler.fitted:  # type: ignore
+                    self.check_threshold_inferred()
+            return self.ensembler(x)  # type: ignore
+        else:
+            return x
+
+    def _classify_outlier(self, scores: torch.Tensor) -> torch.Tensor:
+        """Classify the data as outlier or not.
+
+        Parameters
+        ----------
+        scores
+            Scores to classify. Larger scores indicate more likely outliers.
+
+        Returns
+        -------
+        `torch.Tensor` or ``None``
+        """
+        return scores > self.threshold if self.threshold_inferred else None
```

Comment:
This results in booleans being output in the final `is_outlier` field of the `preds` dictionary. I believe we use 0s and 1s elsewhere in library.